File: app/src/main/java/hexlet/code/games/GcdGame.java
package hexlet.code.games;

import java.util.Random;

public final class GcdGame extends Game {

    private static final int MAX_NUMBER = 10;

    private static int generateNumber() {
        var random = new Random();
        var gcdProtector = random.nextInt(1, MAX_NUMBER + 1);
        return gcdProtector * random.nextInt(1, MAX_NUMBER + 1);
    }

    private static int gcd(int num1, int num2) {
        if (num2 == 0) {
            return num1;
        }

        return gcd(num2, num1 % num2);
    }

    @Override
    public String getRule() {
        return "Find the greatest common divisor of given numbers.";
    }

    @Override
    public GameRound getNewRound() {
        var num1 = generateNumber();
        var num2 = generateNumber();

        var task = "%d %d".formatted(num1, num2);
        var rightAnswer = Integer.toString(gcd(num1, num2));

        return new GameRound(task, rightAnswer);
    }
}

File: app/src/main/kotlin/hexlet/code/GcdGame.kt
package hexlet.code

class GcdGame : Game() {
    override fun rule(): String {
        return "Find the greatest common divisor of given numbers."
    }

    override fun newRound(): GameRound {
        val n1 = generateNumber()
        val n2 = generateNumber()

        val question = question(n1, n2)
        val rightAnswer = rightAnswer(n1, n2)

        return GameRound(question, rightAnswer)
    }

    private fun question(n1: Int, n2: Int): String {
        return "Question: $n1 $n2"
    }

    private fun rightAnswer(n1: Int, n2: Int): String {
        val result = gcd(n1, n2)
        return result.toString()
    }

    private fun gcd(n1: Int, n2: Int): Int {
        if (n2 == 0) return n1

        return gcd(n2, n1 % n2)
    }

    private fun generateNumber(): Int {
        // protect the game from answer '1'
        val gcdProtector = (1..numberLimit).random()

        return gcdProtector * (1..numberLimit).random()
    }

    companion object {
        private const val numberLimit = 10
    }
}
